pdns/dnssecinfra.cc

```diff
@@ -647,8 +625,22 @@ void addTSIG(DNSPacketWriter& pw, TSIGRecordContent* trc, const string& tsigkeyn
   const vector<uint8_t>& signRecord=dw.getRecordBeingWritten();
   toSign.append(&*signRecord.begin(), &*signRecord.end());
 
-  trc->d_mac = calculateHMAC(tsigsecret, toSign, algo);
-  //  d_trc->d_mac[0]++; // sabotage
+  if (algo == TSIG_GSS) {
+    string mac;
+    pdns::GssContext gssctx(tsigkeyname);
+    if (!gssctx.valid()) {
+      L<<Logger::Error<<"Could not add GSS-TSIG signature, valid context not found for keyname '"<<tsigkeyname<<"'"<<endl;
+      return;
+    }
+    if (!gssctx.sign(toSign, mac)) {
+      L<<Logger::Error<<"Could not add GSS-TSIG signature, sign failed"<<endl;
+      return;
+    }
+    trc->d_mac = mac;
+  } else {
+    trc->d_mac = calculateHMAC(tsigsecret, toSign, algo);
+    //  d_trc->d_mac[0]++; // sabotage
```

Comment:
written by @ahupowerdns in 78bcb85857b15320e9cd2b28e8b486a667a1d79f - presumably to make it easy to validate the crypto implementation